File: LeetCodeWP/src/main/java/com/hef/week09/Trie.java
package com.hef.week09;

/**
 * @Date 2021/5/18
 * @Author lifei
 */
public class Trie {

    private Trie[] children;
    private boolean isEnd;

    public Trie() {
        children = new Trie[26];
        isEnd = false;
    }

    public void insert(String word) {
        Trie node = this;
        for (int i=0; i<word.length(); i++) {
            char c = word.charAt(i);
            int index = c - 'a';
            if (node.children[index] == null) {
                node.children[index] = new Trie();
            }
            node = node.children[index];
        }
        node.isEnd = true;
    }

    public boolean search(String word) {
        Trie node = searchPrefix(word);
        return node!=null && node.isEnd;
    }

    public boolean startWith(String prefix) {
        Trie node = searchPrefix(prefix);
        return node!=null;
    }

    private Trie searchPrefix(String prefix) {
        Trie node = this;
        for (int i=0; i<prefix.length(); i++) {
            char c = prefix.charAt(i);
            int index = c-'a';
            if (node.children[index] == null) {
                return null;
            }
            node = node.children[index];
        }
        return node;
    }
}

File: projects/LeetCodeWP/LeetCodePythonWP/week08/Trie.py
#!/usr/bin/python3

class Trie:

    def __init__(self):
        self.root = {}
        self.end_of_word="#"

    def insert(self, word:str):
        node = self.root
        for char in word:
            node = node.setdefault(char, {})
        node[self.end_of_word] = self.end_of_word
    
    def search(self, word: str):
        node = self.root
        for c in word:
            if c not in node:
                return False
            node = node[c]
        return self.end_of_word in node
    
    def startWith(self, word: str):
        node = self.root
        for c in word:
            if c not in node:
                return False
            node = node[c]
        return True

    
